Add support for slash as sysctl separator to Pod securityContext field and to PodSecurityPolicy

mengjiao-liu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Starting from Kubernetes version 1.23, the kubelet supports the use of either `/` or `.`
as separators for sysctl names. 
For example, you can represent the same sysctl name as `kernel.shm_rmid_forced` using a period as the separator, or as `kernel/shm_rmid_forced` using a slash as a separator.
For more sysctl parameter conversion method details, please refer to the page [sysctl.d(5)](https://man7.org/linux/man-pages/man5/sysctl.d.5.html) from the Linux man-pages project.

In 1.25, use relaxed validation everywhere, in other words, Pod SecurityContext and PodSecurityPolicy can supports slash as sysctl separator. And I added the corresponding e2e test.

Ref https://github.com/kubernetes/kubernetes/pull/102393#discussion_r645018213
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #102373

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Pod SecurityContext and PodSecurityPolicy supports slash as sysctl separator.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Commit:
Pod SecurityContext and PodSecurityPolicy supports slash as sysctl separator

mengjiao-liu commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

Commit:
Add test for sysctl support slash as separator

mengjiao-liu commented:
/sig api-machinery

mengjiao-liu commented:
area/test
sig/testing

fedebongio commented:
/remove-sig api-machinery

mengjiao-liu commented:
/assign @smarterclayton @liggitt

liggitt commented:
API bits lgtm, will defer test review to node team

mengjiao-liu commented:
/cc @derekwaynecarr @mrunalp

mengjiao-liu commented:
ping @derekwaynecarr @mrunalp

mengjiao-liu commented:
Request review @sig-node-approvers

mengjiao-liu commented:
Request review @sig-node-approvers

mengjiao-liu commented:
Review review @kubernetes/sig-node-pr-reviews

mengjiao-liu commented:
ping @derekwaynecarr @mrunalp

mrunalp commented on pkg/apis/policy/validation/validation_test.go:
```diff
@@ -807,6 +807,10 @@ func TestIsValidSysctlPattern(t *testing.T) {
+		"a/b/c/d",
+		"a/*",
+		"a/b/*",
+		"a.b/c*",
```
Can we include these two patterns as well?
`
		"a.b/c.d",
		"a/b.c/d",
`

mengjiao-liu replied:
Yes, we can include these two patterns as well.
I'll add them to the test cases.

mengjiao-liu replied:
ping @mrunalp

mengjiao-liu commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

mengjiao-liu commented:
/retest

mengjiao-liu commented:
This PR has been submitted for a long time. We are very near to 1.24 code freeze. The API part has been reviewed and now requires review by sig-node reviewers or approvers. @derekwaynecarr @mrunal @random-liu @dawnchen  @yujuhong @klueska @sjenning

mengjiao-liu commented:
> API bits lgtm, will defer test review to node team

node team has reviewed. Request approve @liggitt

mengjiao-liu commented:
Request code review for the e2e section @smarterclayton @johnbelamaric  @derekwaynecarr @tallclair @yujuhong @ehashman @SergeyKanzhelev

liggitt commented:
are there minimum kernel / container runtime / OS versions required for slash-delimited sysctl support? if so, what is the impact of using a slash-delimited sysctl on a node/kernel/os/container runtime that doesn't support slash-delimited sysctls? is the impact limited to an error running the container, or can it cause problems beyond the scope of that container?

ehashman commented:
/priority important-longterm
/triage accepted

This probably won't make it for 1.24 but we can try to merge early in 1.25.

ehashman commented on test/e2e/common/node/sysctl.go:
```diff
@@ -173,4 +173,56 @@ var _ = SIGDescribe("Sysctls [LinuxOnly] [NodeConformance]", func() {
+	  Description: Pod is created with kernel/shm_rmid_forced sysctl. Support slashes as sysctl separator. The '/' separator is also accepted in place of a '.'
+	  [LinuxOnly]: This test is marked as LinuxOnly since Windows does not support sysctls
+	*/
+	framework.ConformanceIt("should support sysctls with slashes as separator [MinimumKubeletVersion:1.23]", func() {
```
This should not be added directly as a conformance test. Let's start with a cluster e2e and consider promoting after we get good signal from the test.

mengjiao-liu replied:
> Let's start with a cluster e2e

I don't understand what does `cluster e2e test` mean?  Where does it need to be modified?
Do you mean change  `framework.ConformanceIt` to `ginkgo.It` or  test kind?

oomichi replied:
I think @mengjiao-liu understands @ehashman's comment correctly.
A conformance test needs to be stable as `it is stable and runs consistently (e.g., no flakes), and has been running for at least two weeks` from https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements
Then we need to check the stability as `ginkgo.It` for two weeks at least, then the test can promote to a conformance test after that.

mengjiao-liu replied:
Comment addressed. Request review again. Thanks for @ehashman @oomichi

sftim commented:
/sig security

@mengjiao-liu the PR title describes the context, I think. How would you summarize the _change_ that this PR makes?

sftim commented:
Eg, you could 
`/retitle Add support for slash as sysctl separator to Pod securityContext field and to PodSecurityPolicy`

mengjiao-liu commented:
> Eg, you could `/retitle Add support for slash as sysctl separator to Pod securityContext field and to PodSecurityPolicy`

Thanks for your suggestion, it has been changed.

mengjiao-liu commented:
> are there minimum kernel / container runtime / OS versions required for slash-delimited sysctl support?

The documentation on sysctl delimiter conversion is not found on the Internet, so I need to check the linux and sysctl source code.


linux kernel support sysctl:
sysctl appeared in `1.3.57` and is fully supported ever since. 

```
/proc/sys
              This directory (present since 1.3.57) contains a number of
              files and subdirectories corresponding to kernel
              variables.  These variables can be read and in some cases
              modified using the /proc filesystem, and the (deprecated)
              [sysctl(2)](https://man7.org/linux/man-pages/man2/sysctl.2.html) system call.

              String values may be terminated by either '\0' or '\n'.

              Integer and long values may be written either in decimal
              or in hexadecimal notation (e.g., 0x3FFF).  When writing
              multiple integer or long values, these may be separated by
              any of the following whitespace characters: ' ', '\t', or
              '\n'.  Using other separators leads to the error EINVAL.
```
ref: 
- https://man7.org/linux/man-pages/man5/proc.5.html
- https://www.linuxjournal.com/article/2365
- source code: https://mirrors.edge.kernel.org/pub/linux/kernel/v1.3/linux-1.3.57.tar.gz

The [sysctl ](https://en.wikipedia.org/wiki/Sysctl) also uses the conversion of slashes and dot separators,at least 10 years ago.

I found the source code of [ procps-ng Sysctl](https://gitlab.com/procps-ng/procps/-/blob/v3.3.0/sysctl.c#L286), the earliest version is 3.3.0 I can found, and the conversion of . and / has been implemented at this time, at least 10 years ago（Maybe earlier, but I can't find the source code further forward）.


> if so, what is the impact of using a slash-delimited sysctl on a node/kernel/os/container runtime that doesn't support slash-delimited sysctls?
is the impact limited to an error running the container, or can it cause problems beyond the scope of that container?

I think that the kernel versions generally used now are greater than version 1.35.7, and the bottom layer of the kernel supports sysctl including slashes.

As far as my own understanding is concerned, the linux kernel finally needs to convert dots into slashes to find files under /proc/sys, so it must support slashes.

The runc code https://github.com/opencontainers/runc/commit/f7d1401a6998bf2036608e049e858ce174651087 only adds namespace verification to sysctl, and then uses syscall to call sysctl in 2015.

The runc code just adds namespace verification to sysctl, and then calls sysctl.

runc 1.1.0 allow slashes in sysctl names, to better match `sysctl(8)`'s behaviour. (https://github.com/opencontainers/runc/pull/3254, https://github.com/opencontainers/runc/pull/3257)

In order to be compatible with runc which does not support slashes to separate namespaces and keys, I have converted the slashes into dot separators before the kubelet calls runc.
https://github.com/kubernetes/kubernetes/blob/1ea07d482a5ec15afac5ea837585174579fc0f57/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L798

So I don't think it will have any effect.
However, I am not very familiar with the linux kernel source code, maybe someone who is more familiar with this part of the code can help to confirm it again?  @liggitt

enj commented:
/milestone v1.25

pacoxu commented on test/conformance/testdata/conformance.yaml:
```diff
@@ -2195,6 +2195,14 @@
+  description: 'Pod is created with kernel/shm_rmid_forced sysctl. Support slashes
+    as sysctl separator. The ''/'' separator is also accepted in place of a ''.''
+    [LinuxOnly]: This test is marked as LinuxOnly since Windows does not support sysctls'
+  release: v1.23
```
Should we update it to v1.25 for final review?

mengjiao-liu replied:
ref https://github.com/kubernetes/kubernetes/pull/106834#discussion_r875252364 
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements

> it is stable and runs consistently (e.g., no flakes), and has been running for at least two weeks

According to this documentation,I have removed this update from the `test/conformance/testdata/conformance.yaml`. Thanks for the catch.

mengjiao-liu commented:
/retest

pacoxu commented:
/lgtm
move to Done in sig-node board.

mengjiao-liu commented:
Please review again. @liggitt  @lavalamp

liggitt commented on test/e2e/common/node/sysctl.go:
```diff
@@ -175,4 +175,56 @@ var _ = SIGDescribe("Sysctls [LinuxOnly] [NodeConformance]", func() {
+	  Release: v1.23
+	  Testname: Sysctl, test sysctls supports slashes
+	  Description: Pod is created with kernel/shm_rmid_forced sysctl. Support slashes as sysctl separator. The '/' separator is also accepted in place of a '.'
+	  [LinuxOnly]: This test is marked as LinuxOnly since Windows does not support sysctls
```
the test isn't marked `[LinuxOnly]` currently, right?

mengjiao-liu replied:
https://github.com/kubernetes/kubernetes/blob/7e2c02d4c86d73051b6d4f47298d35a7ffb42b6a/test/e2e/common/node/sysctl.go#L35
The top layer of this file has defined `[LinuxOnly]`.

liggitt reviewed (commented):
/approve
for API changes

/hold for one comment on the e2e test

liggitt commented:
/hold cancel

mengjiao-liu commented:
/test pull-kubernetes-e2e-kind
